File: lib/tasks/tree.rb
# frozen_string_literal: true

module Tasks
  # The structural index over the live file: every record (section or task)
  # becomes a Node with its OWN body lines and its children, so callers can ask
  # hierarchy questions the flat Store#items list can't answer — which project a
  # task belongs to, whether a project has a NEXT action, what prose sits under
  # a heading (for body search and link extraction).
  #
  # Hierarchy comes straight from each record's `parent` pointer (no star
  # counting, no block inference). A task node's `item` is the Store item the
  # parser produced for that record; a section node's `item` is nil.
  module Tree
    # Caps task-depth for nesting mutations (capture/move --under); enforcement
    # lives in Store, not Check, so deeper legacy files still validate.
    DEFAULT_MAX_DEPTH = 4

    Node = Struct.new(:title, :line, :level, :item, :body, :children, :parent,
                      keyword_init: true) do
      def task?    = !item.nil?
      def section? = item.nil?

      # Depth-first walk of this node and everything under it.
      def each(&blk)
        return to_enum(:each) unless blk
        yield self
        children.each { |c| c.each(&blk) }
      end

      # The nearest ancestor headline — a task's project (or a subtask's parent
      # task). nil at top level.
      def project = parent

      # The task's project *for grouping/display*: the nearest ancestor headline
      # that is either a section or an OPEN task, climbing PAST closed
      # (DONE/CANCELLED) task ancestors — a closed ancestor is transparent, the
      # same way the outliner hoists open descendants out of a closed parent.
      # Deferred tasks count as open (a deferred project still owns its
      # subtasks). nil when no such ancestor exists (a top-level task, or one
      # whose every task ancestor is closed with no section above it), so those
      # tasks fall out of the Projects view rather than heading a dead group.
      def open_project
        a = parent
        a = a.parent while a && a.task? && !a.item.open?
        a
      end

      # Body prose joined for matching. The record's body is already prose
      # (Format strips nothing — the store stored only prose), so this is a
      # straight join of the node's own lines.
      def body_text = body.join("\n")
    end

    module_function

    # Build the forest of top-level nodes from `records` (as Tasks::Format
    # parsed them, each carrying its physical `line`). `items_by_line` maps a
    # 1-based line number to its Store item, binding each task node to the item
    # the parser produced. Order is DFS pre-order, so a single pass over the
    # records — linking each to its parent by id — reconstructs the tree.
    def build(records, items_by_line)
      roots = []
      by_id = {}
      records.each do |r|
        next if r["type"] == "meta"

        item = items_by_line[r["line"]]
        body = r["body"].nil? || r["body"].empty? ? [] : r["body"].split("\n")
        node = Node.new(
          # A task node's title is the parsed item title; a section's is its
          # own title field.
          title: item ? item.title : r["title"],
          line: r["line"], level: 1,
          item: item, body: body, children: [],
        )
        by_id[r["id"]] = node

        if (pid = r["parent"]) && (parent = by_id[pid])
          node.parent = parent
          node.level = parent.level + 1
          parent.children << node
        else
          roots << node
        end
      end
      roots
    end
  end
end

File: internal/taskquery/tree.go
// Package taskquery is the read model over a store snapshot: the structural
// tree, effective availability, and the selections the read surfaces render.
// It is the Go counterpart of lib/tasks/tree.rb and the read half of
// lib/tasks/task_queries.rb.
//
// Nothing here writes. The whole package answers questions about a snapshot a
// caller is already holding, so two renderings of the same snapshot can never
// disagree.
package taskquery

import (
	"strings"

	"github.com/marcus/tasks/internal/record"
	"github.com/marcus/tasks/internal/store"
)

// Node is one record in the structural index: a section or a task, with its own
// body lines and its children. Hierarchy comes straight from each record's
// `parent` pointer — no star counting, no block inference — so a whole class of
// boundary bugs is structurally absent.
type Node struct {
	Title string
	Line  int
	Level int
	// ID is the record's stable 8-hex id — a task's or a section's. Sections
	// carry no Item, so the id lives on the node itself; it is what the
	// outliner's fold state and selection follow for section rows.
	ID       string
	Item     *store.Item // nil for a section
	Body     []string
	Children []*Node
	Parent   *Node
	// State and Closed are the section lifecycle pair when this node is a
	// section. Tasks carry lifecycle on their Item, sections carry it here
	// so row builders can ask a node directly without a second lookup.
	State     string
	Closed    string
	HasClosed bool
}

// Closed reports whether this section node is closed.
func (n *Node) ClosedView() bool { return n != nil && n.HasClosed }

// Open reports whether this section node is open.
func (n *Node) Open() bool { return n == nil || !n.HasClosed }

// ProjectSection climbs ancestors to the section that is a child of the
// Projects root, or nil when there is none. It is the resolver the task
// detail rail uses to name the project a task belongs to.
func (n *Node) ProjectSection() *Node {
	if n == nil {
		return nil
	}
	current := n.Parent
	for current != nil {
		if current.Section() && current.Parent != nil && isProjectsRoot(current.Parent) {
			return current
		}
		current = current.Parent
	}
	return nil
}

func isProjectsRoot(node *Node) bool {
	if node == nil || node.Parent != nil {
		return false
	}
	return strings.EqualFold(strings.TrimSpace(node.Title), "Projects")
}

// Task reports a node that carries an item.
func (n *Node) Task() bool { return n != nil && n.Item != nil }

// Section reports a node that does not.
func (n *Node) Section() bool { return n != nil && n.Item == nil }

// Project is the nearest ancestor headline: a task's project, or a subtask's
// parent task.
func (n *Node) Project() *Node {
	if n == nil {
		return nil
	}
	return n.Parent
}

// OpenProject is the task's project FOR GROUPING: the nearest ancestor headline
// that is a section or an OPEN task, climbing past closed ancestors. A closed
// ancestor is transparent, the same way the outliner hoists open descendants
// out of a closed parent. Deferred tasks count as open — a deferred project
// still owns its subtasks.
func (n *Node) OpenProject() *Node {
	ancestor := n.Parent
	for ancestor != nil && ancestor.Task() && !isOpen(ancestor.Item.State) {
		ancestor = ancestor.Parent
	}
	return ancestor
}

// Tree is the forest plus the line index every per-item lookup goes through.
type Tree struct {
	Roots       []*Node
	NodesByLine map[int]*Node
}

// BuildTree reconstructs the forest in one pass over the live records, linking
// each to its parent by id. The file is in DFS pre-order, so a single forward
// pass suffices and no ordering is inferred.
func BuildTree(records []record.Record, items []store.Item) Tree {
	itemsByLine := map[int]*store.Item{}
	for index := range items {
		itemsByLine[items[index].Line] = &items[index]
	}
	tree := Tree{NodesByLine: map[int]*Node{}}
	byID := map[string]*Node{}
	for _, parsed := range records {
		if stringOf(parsed, "type") == "meta" {
			continue
		}
		item := itemsByLine[parsed.Line]
		title := stringOf(parsed, "title")
		if item != nil {
			title = item.Title
		}
		body := []string{}
		if text := stringOf(parsed, "body"); text != "" {
			body = strings.Split(text, "\n")
		}
		node := &Node{Title: title, Line: parsed.Line, Level: 1, Item: item, Body: body, Children: []*Node{}}
		if id := stringOf(parsed, "id"); id != "" {
			node.ID = id
			byID[id] = node
		}
		if state := stringOf(parsed, "state"); state != "" {
			node.State = state
			if closed := stringOf(parsed, "closed"); closed != "" {
				node.Closed = closed
				node.HasClosed = true
			}
		}
		if parentID := stringOf(parsed, "parent"); parentID != "" {
			if parent, found := byID[parentID]; found {
				node.Parent = parent
				node.Level = parent.Level + 1
				parent.Children = append(parent.Children, node)
				tree.NodesByLine[node.Line] = node
				continue
			}
		}
		tree.Roots = append(tree.Roots, node)
		tree.NodesByLine[node.Line] = node
	}
	return tree
}

func stringOf(parsed record.Record, key string) string {
	for _, field := range parsed.Fields {
		if field.Key != key {
			continue
		}
		var value string
		if err := unmarshalString(field.Value, &value); err != nil {
			return ""
		}
		return value
	}
	return ""
}
